File: matematica/combinatoria/subconjuntos.cpp
// Gerar subconjuntos

/*
Dado um array de interos A, o algoritmo gera todos os subconjuntos de A.
*/

#include <bits/stdc++.h>
using namespace std;

class Conj {
public:
    vector<int> a, temp;
    vector<vector<int>> subsets;

    Conj(vector<int> a) {
        this->a = a;
    }

    void search(int index) {
        this->subsets.push_back(this->temp);

        for (int i = index; i < this->a.size(); i++) {
            this->temp.push_back(this->a[i]);
            this->search(i + 1);
            this->temp.pop_back();
        }
    }

    void find_subsets() {
        this->search(0);
    }
};

File: matematica/combinatoria/subconjuntos.py
# Gerar subconjuntos
# O(2^N)

# Dado um array de interos A, o algoritmo gera todos os subconjuntos de A.

a = [1, 2, 3]

subsets = []
cur = []

def search(index: int) -> None:
    subsets.append([*cur])

    for i in range(index, len(a)):
        cur.append(a[i])
        search(i + 1)
        cur.pop()

search(0)